pkg/controller/volume/attachdetach/reconciler/reconciler.go

```diff
@@ -93,60 +93,43 @@ func (rc *reconciler) reconciliationLoopFunc() func() {
 			if !rc.desiredStateOfWorld.VolumeExists(
 				attachedVolume.VolumeName, attachedVolume.NodeName) {
 				// Volume exists in actual state of world but not desired
-
-				// Mark desire to detach
-				timeElapsed, err := rc.actualStateOfWorld.MarkDesireToDetach(attachedVolume.VolumeName, attachedVolume.NodeName)
+				timeout, err := rc.actualStateOfWorld.CheckAndMarkDesireToDetach(
```

Comment:
I don't think this change is necessary to fix the issue. The other two changes--"add volume to the list of `volumesToReportAsAttached` only if it has actually gone through a complete/real attach operation," and "re-add volumes to `volumesToReportAsAttached` if a detach operation fails"--are necessary. This change is larger than both of those.

Moreover, it makes the code harder to understand because it is blurring the separation of responsibilities between the reconciler and the ASW data structure. The ASW data structure should be strictly responsible for maintaining state. The reconciler should be responsible for using that state to make decisions about what to do.

It looks like you are trying to fix two issues with this change:

1. If `attachedVolume.MountedByNode` (cached value) is false, but DetachVolume discovers otherwise (real value differs) , it can get stuck and never detach.
2. Prevent a volume from being removed from attached list unless a detach operation is actually kicked off.

If you want to keep this change, I suggest the following instead:

1. Switch the order of the `if !attachedVolume.MountedByNode` and `if timeElapsed > rc.maxWaitForUnmountDuration`checks (so that `if timeElapsed > rc.maxWaitForUnmountDuration` is checked first) instead of moving the check inside of `MarkDesireToDetach` and overloading its behavior.
2. Split `MarkDesireToDetach` into two separate methods `SetDetachRequestedTime` (which sets the `detachRequestedTime` to now) and `RemoveVolumeFromReportAsAttached`. Replace the `MarkDesireToDetach` call with `SetDetachRequestedTime`. Remove the `UpdateNodeStatuses` call. Replace the two `DetachVolume` calls with a call (helper function in this file) that will call `RemoveVolumeFromReportAsAttached` then `UpdateNodeStatuses` and then `DetachVolume`.

I don't agree that the second point is a cause for concern, I think that behavior is safer.